doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -522,28 +505,26 @@ the conversion in hypotheses :n:`{+ @ident}`.
    information about the constants it encounters and the unfolding decisions it
    makes.
 
-.. tacn:: red
-   :name: red
+.. tacn:: red {? @occurrences }
 
-   This tactic applies to a goal that has the form::
+   :math:`\beta`:math:`\iota`:math:`\zeta`-reduces `T`, if possible, in the selected
+   hypotheses and/or the goal, which must have the form:
 
-     forall (x:T1) ... (xk:Tk), T
+     :n:`forall @open_binders, T`
 
-   with :g:`T` :math:`\beta`:math:`\iota`:math:`\zeta`-reducing to :g:`c t`:sub:`1` :g:`... t`:sub:`n` and :g:`c` a
+   to :g:`c t`:sub:`1` :g:`… t`:sub:`n` where :g:`c` is a
    constant. If :g:`c` is transparent then it replaces :g:`c` with its
-   definition (say :g:`t`) and then reduces
-   :g:`(t t`:sub:`1` :g:`... t`:sub:`n` :g:`)` according to :math:`\beta`:math:`\iota`:math:`\zeta`-reduction rules.
+   definition and reduces again until no further reduction is possible.
 
-.. exn:: Not reducible.
-   :undocumented:
+   .. exn:: Not reducible.
+      :undocumented:
 
-.. exn:: No head constant to reduce.
-   :undocumented:
+   .. exn:: No head constant to reduce.
+      :undocumented:
 
-.. tacn:: hnf
-   :name: hnf
+.. tacn:: hnf {? @occurrences }
 
-   This tactic applies to any goal. It replaces the current goal with its
+   Replaces the current goal with its
    head normal form according to the :math:`\beta`:math:`\delta`:math:`\iota`:math:`\zeta`-reduction rules, i.e. it
    reduces the head of the goal until it becomes a product or an
    irreducible term. All inner :math:`\beta`:math:`\iota`-redexes are also reduced.
```

Comment:
How does `hnf` differ from `simpl`?

> Is it worth talking about fixpoint refolding here ... It seems at the very least describing to what extent arguments are reduced (whether it's like `lazy` or `cbv` or neither) before being substituted.

I've no idea whether it's worth mentioning or what to say.  If you can provide a couple short points to me, I'll try to add them in.

Reply:
Once the term is in weak-head normal form, `simpl` and `cbn` recurse into subterms while `hnf` does not.  I believe `hnf` will do more unfolding than `simpl` (it will unfold definitions at the head unless forbidden by things like `simpl never`) and I'm not sure if fixpoint refolding applies to the head constant (does `intros; hnf` on `forall xs, fold_right and True xs` behave differently from `intros; simpl`?)

Fixpoint refolding is why `simpl` on `S x + y` gives `S (x + y)` rather than `S ((fix add n m ... ) x y)` which is what `cbv` gives.  It's probably the feature that is the biggest part of giving `simpl` and `cbn` the appearance of "simplifying" rather than "computing".  (The other biggest part is figuring out what to unfold and what to not unfold in the first place.)